### PR TITLE
Fix AS::Callbacks raising an error when `:run` callback is defined.

### LINKED ISSUE #19405: Defining :run callbacks breaks whole callbacks infrastructure
When you define `:run` callbacks, anything that calls any callbacks breaks with quite vague error message:

```
/home/katafrakt/.rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/activesupport-4.2.0/lib/active_support/core_ext/class/attribute.rb:106:in `block (2 levels) in class_attribute': wrong number of arguments (1 for 0) (ArgumentError)
	from /home/katafrakt/.rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/activesupport-4.2.0/lib/active_support/callbacks.rb:734:in `_run_run_callbacks'
	from /home/katafrakt/.rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/activesupport-4.2.0/lib/active_support/callbacks.rb:81:in `run_callbacks'
```

This behaviour can be easily reproduced with following script:

```ruby
gem 'activesupport', '4.2.0'
require 'active_support/all'

class Machine
	include ActiveSupport::Callbacks
	define_callbacks :run

	def start
		run_callbacks(:run) do
			puts "running"
		end
	end
end

m = Machine.new
m.start
```

This is caused by how `define callbacks` method works. In this case it creates class attribute `_run_callbacks` and defines method `_run_run_callbacks` that calls internal `_run_callbacks` method, which is now overriden by class attribute. In fact, running any callbacks (not only `run`) will result in error message above.

```ruby
      def define_callbacks(*names)
        options = names.extract_options!

        names.each do |name|
          class_attribute "_#{name}_callbacks"
          set_callbacks name, CallbackChain.new(name, options)

          module_eval <<-RUBY, __FILE__, __LINE__ + 1
            def _run_#{name}_callbacks(&block)
              _run_callbacks(_#{name}_callbacks, &block)
            end
          RUBY
        end
      end
```

It works with version 3.2.15, I'm not sure about intermediate versions. Master is also affected.

**Proposed solution**

Raise error with clear message when user tries to define `:run` callback. I can PR this unless someone comes up with better idea.

### PR DESCRIPTION
**Pull request by tgxworld**
Fixes: https://github.com/rails/rails/issues/19405.

Based on the reverted commit, I ran the [script](https://github.com/rails/rails/pull/17093) that was checking for object allocations but didn't see any difference even after reverting the commit.

``` bash
rails (fix_activesupport_callbacks_clash_on_run) $ ruby /home/guoxiang/work/rails-dev-box/rails/ob.rb > after.txt
rails (fix_activesupport_callbacks_clash_on_run) $ git co master
Switched to branch 'master'
Your branch is ahead of 'origin/master' by 1732 commits.
  (use "git push" to publish your local commits)
rails (master) $ ruby /home/guoxiang/work/rails-dev-box/rails/ob.rb > before.txt
rails (master) $ git diff before.txt after.txt
```

Similarly there isn't any significant difference in performance as well based on the [benchmarks](https://gist.github.com/phiggins/e46e51dcc7edb45b5f98) used in the reverted commit. 

``` bash
Calculating -------------------------------------
                find     1.729k i/100ms
          initialize     9.850k i/100ms
              create   295.000  i/100ms
-------------------------------------------------
                find     20.690k (± 0.8%) i/s -    620.711k
          initialize    157.055k (± 1.7%) i/s -      4.708M
              create      3.059k (± 0.7%) i/s -     91.745k
guoxiang@guoxiang-GS60-2PC-Ghost rails (fix_activesupport_callbacks_clash_on_run) $ git co master
Switched to branch 'master'
Your branch is ahead of 'origin/master' by 1732 commits.
  (use "git push" to publish your local commits)
guoxiang@guoxiang-GS60-2PC-Ghost rails (master) $ ruby /home/guoxiang/work/rails-dev-box/rails/bench.rb
Calculating -------------------------------------
                find     1.735k i/100ms
          initialize    10.451k i/100ms
              create   296.000  i/100ms
-------------------------------------------------
                find     20.264k (± 1.0%) i/s -    608.985k
          initialize    152.585k (± 1.7%) i/s -      4.578M
              create      3.057k (± 0.8%) i/s -     91.760k
```

cc/ @phiggins @tenderlove @rafaelfranca

### CODE REVIEW

File: activesupport/lib/active_support/callbacks.rb [hunk 1]
```diff
@@ -79,14 +79,10 @@ module Callbacks
...
-      send "_run_#{kind}_callbacks", &block
-    end
-
-    private
+      callbacks = send("_#{kind}_callbacks")
 
-    def _run_callbacks(callbacks, &block)
       if callbacks.empty?
-        block.call if block
+        yield if block_given?
```
Comment (tgxworld) [inline activesupport/lib/active_support/callbacks.rb]: Previously it was optimized to use `block.call`. However, the benchmark script used in https://github.com/rails/rails/pull/17093 did not actually pass a block to the method being called. It was just calling the method with no block.

Stealing benchmark results from https://www.omniref.com/ruby/2.2.0/symbols/Proc/yield#annotation=4088660&line=723, I believe we should be using yield.

**Comment by tgxworld** [comment]
Not sure why Travis isn't showing up here. https://travis-ci.org/rails/rails/builds/55342228

**Comment by tgxworld** [comment]
Any updates on this? :smile:

**Comment by rafaelfranca** [comment]
Backported at fe664e8

**Pull request merged**